### PR TITLE
Create crayons mobile drawer components

### PR DESCRIPTION
**Pull request by aitchiss**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This PR introduces the design systems components: `MobileDrawer` and `MobileDrawerNavigation` which will be used as part of https://github.com/forem/rfcs/pull/236 to update the home feed UX for users on mobile devices.

These are currently used in Storybook only, and there are no app changes as part of this PR. I wouldn't consider these components "final" in this PR - it's just some initial groundwork, and I think we'll see more opportunities to improve as we get to actually implementing them in the app.

### `MobileDrawer`

This is essentially a variation on a modal, with the key difference being that this animates in from the bottom of the screen, and only expands in height to accommodate the given content.

Other functionality is pretty much the same as the modals we use throughout the app including
- Transferring and trapping focus when opening/closing to ensure accessibility
- Escape key closes the drawer (we don't anticipate this component being used on desktop, but good practice to include this anyway)
- Tapping on the overlay closes the drawer

https://user-images.githubusercontent.com/20773163/129346214-f1d5ae42-1cb8-43a4-a070-2617b0652a69.mp4

### `MobileDrawerNavigation`

This component is intended to be used at the top of a given page, and is responsible for displaying the page heading, and a `MobileDrawer` that displays a list of navigation options, indicating which is the current page. For https://github.com/forem/rfcs/pull/23 this will be the mobile-alternative to the feed tabs, and we can expect to drop this feature in other places in the future.

https://user-images.githubusercontent.com/20773163/129346232-8fe58695-aa28-4eb4-b1c2-f870f7467ea1.mp4

## Related Tickets & Documents

Groundwork for https://github.com/forem/rfcs/pull/23

## QA Instructions, Screenshots, Recordings

- Run storybook with `yarn storybook`
- Visit the sections for `Components > MobileDrawer` and `App Components > MobileNavigationDrawer` and explore both the Preact and HTML variants - have a play about with the functionality 🙂  (I suggest setting your devtools to show a mobile viewport to get a truer sense of how this will look and feel)
- Have a look at the notes section for both components and let me know if there's anything you think we could usefully add or change

### UI accessibility concerns?

The `MobileDrawer` component has been created with the [accessibility requirements for dialogs](https://www.digitala11y.com/dialog-role/) in mind, including appropriate labelling and focus management. The "appearance" animation will be reset along with any other animations as part of our `reset.scss` styles for users who have `prefers-reduced-motion` set.

The `MobileDrawerNavigation` component uses the `MobileDrawer` component to display the navigation options. The component requires props that ensure the heading level can be changed, allowing for flexibility of use without degrading accessibility. It also employs `aria-current` to make sure the currently active page is announced to screen reader users, and requires a `navigationTitle` prop which is used to correctly label interactive controls relating to the dialog and navigation elements.

Axe component level tests have been added for both.

## Added/updated tests?

- [X] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://admin.forem.com/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [X] This change does not need to be communicated, and this is why not: I think we're better waiting to communicate this when the components are actually implemented/used in the app itself

## [optional] Are there any post deployment tasks we need to perform?

N/A

## [optional] What gif best describes this PR or how it makes you feel?

Moira Schitt trying to get reception on her mobile

### CODE REVIEW

**Review by msarit — APPROVED** [review]
Both Storybook components work great! 👍🏾 

I had a question @aitchiss: For the `MobileDrawerNavigation`, choosing a link seems to cause a refresh - is this intentional?

**Comment by aitchiss** [comment]
> I had a question @aitchiss: For the MobileDrawerNavigation, choosing a link seems to cause a refresh - is this intentional?

It is @msarit! Or well sort of. It's expected for now. 

When we actually implement this for the home feed in the app, we'll want to maintain the feel that we have the current `select` we use on mobile, but that will likely involve some InstantClick magic ✨  that we don't have in storybook

File: app/javascript/crayons/MobileDrawer/__stories__/drawers.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+button, link) to make sure focus may be transferred to the new content.
+
+### MobileDrawer accessibility
+
+The MobileDrawer is essentially a
+[modal dialog](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/dialog_role).
+A `title` should be passed in props to properly label the dialog, and at least
+one interactive item should be present in the inner content.
+
+The MobileDrawer component utilises the `focus-trap` library to ensure that:
```
Comment (fdoxyz) [inline app/javascript/crayons/MobileDrawer/__stories__/drawers.md:17]: Thank you for these docs! Incredibly helpful to understand how these components work 🙏

**Review by fdoxyz — APPROVED** [review]
These look and work great on Storybook, thank you @aitchiss! 🙌 

> I wouldn't consider these components "final" in this PR - it's just some initial groundwork, and I think we'll see more opportunities to improve as we get to actually implementing them in the app.

I agree with the sentiment and think it's a great step already to have this component available. I do have one idea that may be useful in this or a future PR:

Tablets are technically mobile too, but a full-width view of them is a bit jarring sometimes. For example navigation links would be aligned to the left and a large portion of the drawer would be empty, or the links would be centered and a large portion of the drawer on each end remains unused/empty.

Screen Shot 2021-08-13 at 10 34 34

The suggestion here would be to limit the width when the sizes are larger than phones, maybe a percentage padding/margin on each side or a max-width limit of some sort? 

I'm not sure what the best approach would be, because each platform has their own "native look and feel", but our crayons look and feel doesn't have to exactly match each of them... Here's a quick mock example of what I mean.

Screen Shot 2021-08-13 at 10 42 20

Despite the iOS modal floats right in the middle of the screen (Android would maybe look differently) it's totally okay for us to stay at the bottom (of course this is a "Drawer" 😄 ). By adjusting the width it might look better in tablets.

Anyways enough rambling/thinking out loud, this is great and thanks again @aitchiss!

**Review by rhymes — APPROVED** [review]
This is just awesome! Great work @aitchiss!

**Comment by aitchiss** [comment]
@fdoxyz that's a great point about wider mobile devices! 

The `select` navigation this will initially be used to replace is currently only used on the very smallest screen size (even landscape small mobile phone screens use the desktop navigation tabs), but I think with your suggestion it could easily be used on slightly larger mobile viewports.

I'll set a max-width on it for now, and that'll give us a working example for folks to consider if we want to expand the usage of the new nav component to slightly larger mobile screens 🙌

**Pull request merged**